text/0000-buildpack-registry-search-api.md

```diff
@@ -0,0 +1,208 @@
+# Meta
+[meta]: #meta
+- Name: Buildpack Registry Search API
+- Start Date: 2020-11-11
+- Author(s): elbandito
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+This is a proposal for a service API that enables clients to search for buildpacks.
+
+# Definitions
+[definitions]: #definitions
+
+**API** - Application Programmer Interface
+
+**Buildpack Registry** - the CNB centeralized [registry](https://github.com/buildpacks/registry-index) that contains metadata for all the community supported buildpacks 
+
+**Service** - an independent web process running in the cloud that provides RESTful endpoints
+
+**Client** - represents a physical end-user or software interface e.g. CLI, web UI, etc.
+ 
+# Motivation
+[motivation]: #motivation
+
+As the CNB project makes progress towards supporting a centralized Buildpack Registry, I anticipate the need to quickly search for and identify specific buildpacks.  To support this, I propose the creation of an API search service that customers could use directly and/or indirectly via a web dashboard or CLI.
+
+- What use cases does it support?  
+  - Customers looking to quickly indentify usefulful buildpacks to build their applications.
+- What is the expected outcome?  
+  - Customers will be able to quickly find relevant buildpacks for their use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+An external API service that exposes endpoints for retrieving buildpack metadata.  Initially, I envision this service providing a few GET endpoints to support plain text query searches, and individual buildpack data retrieval. 
+
+### Supported Endpoints
+
+- **GET /search?query=text**
+
+  Retrievs all the buildpacks that satisfy the search query.  If the response contains more than 20 buildpacks, pagination will be used.  For example:
+  ```
+  $ curl https://registry.buildpacks.io/search?query=projectriff
+  ```
+  ```json
+  [
+    {
+      "ns":"projectriff",
+      "name":"command-function",
+      "latestVersion":"1.4.1",
+      "addr":"gcr.io/projectriff/command-function@sha256:99f9054abb73635a9b251b61d3627a8ff86508c767f9d691c426d45e8758596f",
+      "versions": {
+        "1.4.1": {
+          "link": "https://registry.buildpacks.io/buildpacks/projectriff/command-function/1.4.1"
+        },
+        "1.3.9": {
+          "link": "https://registry.buildpacks.io/buildpacks/projectriff/command-function/1.3.9"
+        }
```

Comment:
It would be great if the data model was more alike to that of `/buildpacks/:ns/:name/:version` so that there is only one data structure to represent a "buildpack" when consuming the API. 

For example:

```suggestion
     "latest": {
          "ns":"projectriff",
          "name":"command-function",
          "version":"1.4.1",
          "license": "MIT",
          "description": "...",
          "addr":"gcr.io/projectriff/command-function@sha256:99f9054abb73635a9b251b61d3627a8ff86508c767f9d691c426d45e8758596f"
      },
      "versions": {
        "1.4.1": {
          "link": "https://registry.buildpacks.io/buildpacks/projectriff/command-function/1.4.1"
        },
        "1.3.9": {
          "link": "https://registry.buildpacks.io/buildpacks/projectriff/command-function/1.3.9"
        }
```

Reply:
Good suggestion, @jromero